src/R2_alg.jl

```diff
@@ -117,39 +233,38 @@ function R2(
   end
 
   # initialize parameters
-  xk = copy(x0)
-  hk = h(xk[selected])
+  hk = @views h(xk[selected])
   if hk == Inf
     verbose > 0 && @info "R2: finding initial guess where nonsmooth term is finite"
     prox!(xk, h, x0, one(eltype(x0)))
-    hk = h(xk[selected])
+    hk = @views h(xk[selected])
     hk < Inf || error("prox computation must be erroneous")
     verbose > 0 && @debug "R2: found point where h has value" hk
   end
   hk == -Inf && error("nonsmooth term is not proper")
 
-  xkn = similar(xk)
-  s = zero(xk)
-  ψ = has_bnds ? shifted(h, xk, l_bound - xk, u_bound - xk, selected) : shifted(h, xk)
+  if has_bnds
+    @. l_bound_m_x = l_bound - xk
+    @. u_bound_m_x = u_bound - xk
+    ψ = shifted(h, xk, l_bound_m_x, u_bound_m_x, selected)
+  else
+    ψ = shifted(h, xk)
```

Comment:
There are also allocations here, right? We can think about it in a separate PR, but we should have an issue to track them.

Reply:
Yes, I forgot this. Can we leave it as it is for this PR, and maybe create a function `shifted!` in `ShiftedProximalOperators` to allow us to preallocate the ShiftedProximalOperator structure?